DM-51670: Cleanly separate parent and object catalogs

fred3m opened the pull request.

fred3m committed:
Cleanly separate parent and object catalogs

taranu commented on python/lsst/meas/extensions/scarlet/footprint.py:
```diff
@@ -148,12 +148,10 @@ def scarletModelToHeavy(
     spans = SpanSet.fromMask(valid)
 
     # Add the location of the source to the peak catalog
```
Move comment down.

taranu commented on python/lsst/meas/extensions/scarlet/io/utils.py:
```diff
@@ -195,43 +197,33 @@ def updateCatalogFootprints(
-                parentRecord.getFootprint(), imageForRedistribution.mask
-            )
-
+    blend_items = list(modelData.blends.items())
```
Was this for debugging? It doesn't look like it's necessary.

fred3m replied:
No. This is because the blends can be removed from inside the loop, so I need to generate the full set of blend, parentId pairs or removing items in the loop will cause unexpected behavior.

taranu commented on python/lsst/meas/extensions/scarlet/io/utils.py:
```diff
@@ -195,43 +197,33 @@ def updateCatalogFootprints(
         )
 
+        if removeScarletData:
+            if parentId in modelData.blends:
```
Not that it matters here, but FYI you can do `modelData.blends.pop(parentId, None)`.

taranu commented on python/lsst/meas/extensions/scarlet/io/utils.py:
```diff
@@ -195,43 +197,33 @@ def updateCatalogFootprints(
 def buildMonochromaticObservation(
     modelPsf: np.ndarray,
     observedPsf: np.ndarray,
+    scarletBox: Box,
```
Needs docstring

taranu commented on python/lsst/meas/extensions/scarlet/io/utils.py:
```diff
@@ -326,21 +310,19 @@ def calculateFootprintCoverage(footprint: afwFootprint, maskImage: MaskX) -> np.
+        Persistable data for a single blend or hierarchical blend.
     bandIndex :
         The number of the band to extract.
     parent :
```
Needs to be deleted (also removeScarletData)

taranu commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -685,17 +685,19 @@ class ScarletDeblendConfig(pexConfig.Config):
-        "The key is the name of the column for the parent record, "
-        "the value is the name of the column to use for the child.",
+            "This is no longer used since the object and parent catalogs contain different columns.",
+        deprecated="This field is deprecated along with `morphImage` and will be removed in v30.0.",
```
after v30 (we're already on v29).

taranu commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -887,6 +887,18 @@ def _addParentSchemaKeys(self, schema: afwTable.Schema):
+            "deblend_nComponents",
+            type=np.int32,
+            doc="Number of components in a ScarletLiteSource. "
+            "If `config.version != 'lite'`then "
```
Space before then, but also, is version ever _not_ lite?

fred3m replied:
No, but the config version is deprecated and can be removed. I need to make a ticket for this.

taranu commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -1060,29 +1069,35 @@ def deblend(
+            to (potentially) deblend. If a parent is subdivided into
+            multiple deconvolved parents, the deconvolved parents are
+            added to this catalog in place.
+        objectCatalog :
```
Out of order (did you actually want objectCatalog to come first?)

fred3m replied:
It doesn't matter, so I switched the docstring order.

taranu commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -1762,14 +1800,14 @@ def _buildIntersectingFootprints(
         self,
         parentId: int,
-        blendCatalog: afwTable.SourceCatalog,
+        parentCatalog: afwTable.SourceCatalog,
```
Update docstring names too.

I strongly suggest adding a `pyproject.toml` to this package with `[tool.numpydoc_validation]` enabled (see meas_extensions_multiprofit for an example, or probably daf_butler and others).

fred3m replied:
Strangely the docstring already had `parentCatalog`. I think that originally it took a `parentCatalog`, then it was switched to `parentCatalog` but I forgot to rename the docstring. So now it's actually correct again.

fred3m replied:
I've been meaning to look into adding a pyproject.toml but other things keep coming up. I created DM-53432 to make the addition.

taranu commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -1788,16 +1826,16 @@ def _addBlendRecord(
+        objectCatalog: afwTable.SourceCatalog,
         scarletSource: scl.Source,
         chi2: scl.Image,
     ):
```
Update docstring kwarg names

taranu commented on python/lsst/meas/extensions/scarlet/scarletDeblendTask.py:
```diff
@@ -1854,11 +1886,71 @@ def _addDeblendedSource(
+        src.set("deblend_peak_center_y", peak["i_y"])
+        src.set("deblend_peakId", peak["id"])
+
+        # Isololated sources are not modeled, so they don't have components
```
Isolated

taranu reviewed (commented):
Minor corrections/suggestions, as usual.

fred3m committed:
Mark all skipped flags as deblend_skipped_<reason>

This is a requested feature by end users to make it easier to
understand which columns are associated with reasons that blends
were skipped.
